core/src/hir/attrs.rs

```diff
@@ -673,17 +676,31 @@ impl Attrs {
                                 .push(LoweringError::Other(format!("{name} must return a value")));
                         }
                     }
-                    SpecialMethod::AddAssign => {
-                        check_param_count("AddAssign", 1, errors);
-                    }
-                    SpecialMethod::SubAssign => {
-                        check_param_count("SubAssign", 1, errors);
-                    }
-                    SpecialMethod::MulAssign => {
-                        check_param_count("MulAssign", 1, errors);
-                    }
-                    SpecialMethod::DivAssign => {
-                        check_param_count("DivAssign", 1, errors);
+                    e @ (SpecialMethod::AddAssign
+                    | SpecialMethod::SubAssign
+                    | SpecialMethod::MulAssign
+                    | SpecialMethod::DivAssign) => {
+                        let name = match e {
+                            SpecialMethod::AddAssign => "AddAssign",
+                            SpecialMethod::SubAssign => "SubAssign",
+                            SpecialMethod::MulAssign => "MulAssign",
+                            SpecialMethod::DivAssign => "DivAssign",
+                            _ => unreachable!(),
+                        };
+                        check_param_count(name, 1, errors);
+                        check_self_param(name, true, errors);
+                        if let Some(self_param) = &method.param_self {
+                            if self_param.ty.is_immutably_borrowed() {
+                                errors.push(LoweringError::Other(format!(
+                                    "{name} must take self by mutable reference"
```

Comment:
ideally the error can detect if it's on a struct or enum and tell you to just use add